File: smart_assessment_app/lib/pdf/extract.ts
import { spawnSync } from "child_process";
import { existsSync } from "fs";
import path from "path";

const MAX_PAGES = 25;

export type PdfExtractionResult = {
  pageCount: number;
  usedPages: number;
  chunks: Array<{ page: number; text: string }>;
  hasEnoughText: boolean;
  truncated: boolean;
};

export function extractPdfText(filePath: string): PdfExtractionResult {
  const scriptPath = path.join(process.cwd(), "lib", "pdf", "extract.py");

  if (!existsSync(scriptPath)) {
    throw new Error("PDF extraction script is missing.");
  }

  const result = spawnSync("python", [scriptPath, filePath], {
    encoding: "utf8",
  });

  if (result.error) {
    throw new Error(`PDF extraction failed: ${result.error.message}`);
  }

  if (result.status !== 0) {
    const stderr = result.stderr?.trim() || "Unknown Python error";
    throw new Error(`PDF extraction failed: ${stderr}`);
  }

  const parsed = JSON.parse(result.stdout || "{}") as {
    pageCount?: number;
    usedPages?: number;
    chunks?: Array<{ page: number; text: string }>;
  };

  const chunks = Array.isArray(parsed.chunks) ? parsed.chunks : [];
  const text = chunks.map((chunk) => chunk.text).join("\n\n").trim();
  const pageCount = Number(parsed.pageCount ?? chunks.length ?? 0);
  const usedPages = Number(parsed.usedPages ?? Math.min(pageCount, MAX_PAGES));

  return {
    pageCount,
    usedPages,
    chunks,
    hasEnoughText: text.length > 200,
    truncated: pageCount > MAX_PAGES,
  };
}

File: smart_assessment_app/lib/pdf/extract.py
import json
import sys

import fitz

MAX_PAGES = 25


def extract_text(path: str) -> dict:
    doc = fitz.open(path)
    page_count = len(doc)
    pages = min(page_count, MAX_PAGES)
    chunks = []

    for i in range(pages):
        text = doc[i].get_text("text").strip()
        if text:
            chunks.append({"page": i + 1, "text": text})

    doc.close()
    return {
        "pageCount": page_count,
        "usedPages": pages,
        "chunks": chunks,
    }


if __name__ == "__main__":
    if len(sys.argv) < 2:
        raise SystemExit("Missing file path")
    print(json.dumps(extract_text(sys.argv[1])))
